[TECH] [Back] Isoler et factoriser les locales (PIX-18903)

lego-technix opened the pull request:
## 🔆 Problème

Côté Back, dans Pix API, les notions de _locales_ et de _languages_ sont complètement mélangées, et les déclarations des locales et des langues sont éparpillées entre de nombreux services, modules, fichiers et avec des notations différentes, parfois canoniques (par exemple `fr-FR`) et parfois tout en minuscules (par exemple `fr-fr`) : 
* [api/src/shared/domain/services/locale-service.js](https://github.com/1024pix/pix/blob/dev/api/src/shared/domain/services/locale-service.js)
* [api/src/shared/domain/services/language-service.js](https://github.com/1024pix/pix/blob/dev/api/src/shared/domain/services/language-service.js)
* [api/src/shared/domain/constants.js](https://github.com/1024pix/pix/blob/dev/api/src/shared/domain/constants.js)

## ⛱️ Proposition

La stratégie globale de cette PR est de : 
* Tout isoler dans [api/src/shared/domain/services/locale-service.js](https://github.com/1024pix/pix/blob/dev/api/src/shared/domain/services/locale-service.js),
* Définir une nouvelle notion qui a toujours existé mais qui n’était pas nommée jusque là : la notion de _challengeLocales_, qui correspond aux locales des épreuves du référentiel et qui ne sont pas au format canonique (certaines épreuves ont `fr-fr`),
* Conformément à l’[ADR Logique et stratégie de gestion des paramètres régionaux et des langues (locales & languages)](https://github.com/1024pix/pix/blob/dev/docs/adr/0040-locales-languages.md), commencer à faire disparaitre progressivement la notion de _language_ au profit de la notion unique de _locale_.

Pour éviter un big bang et des PR gigantesques, la disparition totale de la notion de language ne pourra se faire que par étapes.

Plus en détail ce que fait cette PR : 

* Isoler les constantes liées aux locales et langues dans [api/src/shared/domain/services/locale-service.js](https://github.com/1024pix/pix/blob/dev/api/src/shared/domain/services/locale-service.js) : 
  * `LOCALE.XXX` (avec renommage en `CHALLENGE_LOCALES`)
  * `DEFAULT_LOCALE`
  * `SUPPORTED_LOCALES`
* Déplacer le contenu de [api/src/shared/domain/services/language-service.js](https://github.com/1024pix/pix/blob/dev/api/src/shared/domain/services/language-service.js) (souvent injecté en tant que _languageService_) dans le _localeService_,
* Déplacer le code `languageService.isLanguageAvailableForV3Certification` vers le modèle afférent `CertificationCourse`.

## 🌊 Remarques

Cette PR fait partie d’un ensemble de PR sur le sujet de la remise à plat des locales et des langues. De nombreuses PR ont été faites sur les front apps : 
* https://github.com/1024pix/pix/pull/13045
* https://github.com/1024pix/pix/pull/13018
* https://github.com/1024pix/pix/pull/12982
* https://github.com/1024pix/pix/pull/12968
* https://github.com/1024pix/pix/pull/12959
* https://github.com/1024pix/pix/pull/12954
* https://github.com/1024pix/pix/pull/12933
* https://github.com/1024pix/pix/pull/12896


## 🏄 Pour tester

* Relire les modifications faites  au niveau du code pour s’assurer qu’elles préservent bien les logiques
* Vérifier que la création d’utilisateur fonctionne bien et suivant la bonne `lang` (cette PR ne touche pas à la _locale_ de l’utilisateur)
* Vérifier que les envois d’email (mot de passe perdu, invitations, etc.) se font toujours bien et suivant les bonnes locales/langues
* Vérifier que la fonction `CertificationCourse.isLanguageAvailableForV3Certification` fonctionne bien

lego-technix committed:
refactor(api): rename localeFromCookie→locale

since usecase should not be aware of infrastructure matters

pix-bot-github commented:
Choisir les applications à déployer :

- [x] [API](https://api-pr13017.review.pix.fr/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr13017/environment) <!-- pix-api-review -->
- [x] [App (.fr)](https://app-pr13017.review.pix.fr) – [App (.org)](https://app-pr13017.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-app-review-pr13017/environment) <!-- pix-app-review -->
- [x] [Orga (.fr)](https://orga-pr13017.review.pix.fr) – [Orga (.org)](https://orga-pr13017.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-orga-review-pr13017/environment) <!-- pix-orga-review -->
- [x] [Certif (.fr)](https://certif-pr13017.review.pix.fr) – [Certif (.org)](https://certif-pr13017.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-certif-review-pr13017/environment) <!-- pix-certif-review -->
- [ ] [Junior](https://junior-pr13017.review.pix.fr) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-junior-review-pr13017/environment) <!-- pix-junior-review -->
- [ ] [Admin](https://admin-pr13017.review.pix.fr) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-admin-review-pr13017/environment) <!-- pix-admin-review -->
- [ ] [API MaDDo](https://pix-api-maddo-review-pr13017.osc-fr1.scalingo.io/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-maddo-review-pr13017/environment) <!-- pix-api-maddo-review -->
- [ ] [Audit Logger](https://pix-audit-logger-review-pr13017.osc-fr1.scalingo.io/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr13017/environment) <!-- pix-audit-logger-review -->

> [!IMPORTANT]
> N'oubliez pas de déployer l'API pour pouvoir accéder aux fronts et/ou à l’API MaDDo.

lego-technix committed:
refactor(api): isolate SUPPORTED_LOCALES in localeService

lego-technix committed:
refactor(api): isolate DEFAULT_LOCALE in localeService

lego-technix committed:
refactor(api): relocate isLanguageAvailableForV3Certification code from languageService to the CertificationCourse model where it belongs

And it's GOOD to have some specific locale/language names in the model because
those locales/languages are specific to the model and not generic so as to be
located in a service.

lego-technix committed:
refactor(api): replace localeService.assertLanguageAvailability by localeService.coerceLanguage

lego-technix committed:
refactor(api): replace languageService.LANGUAGES_CODE and custom defaultLocale by localeService.getChallengeLocales and localeService.getDefaultChallengeLocale

lego-technix committed:
test(api): replace languageService.LANGUAGES_CODE by locale names

And it's GOOD to have some specific locale/language names in these tests because
those locales/languages are specific to the tests and not generic so as to be
located in a service.

lego-technix committed:
refactor(api): replace languageService.LANGUAGES_CODE by locale names

And it's GOOD to have some specific locale/language names in these seeds because
those locales/languages are specific to these seeds and not generic so as to be
located in a service.

lego-technix committed:
refactor(api): remove remnants of the languageService

lego-technix committed:
refactor(api): in Joi validation replace languageService.LANGUAGES_CODE by localeService.getSupportedLanguages

lego-technix committed:
feat(api): add localeService.getDefaultChallengeLocale

lego-technix committed:
feat(api): add ENGLISH_SPOKEN, FRENCH_FRANCE, FRENCH_SPOKEN, DUTCH_SPOKEN, SPANISH_SPOKEN and getChallengeLocales in localeService

lego-technix committed:
refactor(api): use ENGLISH_SPOKEN, FRENCH_FRANCE, FRENCH_SPOKEN, DUTCH_SPOKEN, SPANISH_SPOKEN and getChallengeLocales from localeService

bpetetot commented on api/src/certification/shared/domain/models/CertificationCourse.js:
```diff
@@ -336,8 +338,10 @@ class CertificationCourse {
-  static isLanguageAvailableForV3Certification(languageService, candidateLanguage) {
-    return languageService.isLanguageAvailableForV3Certification(candidateLanguage);
+  static isLanguageAvailableForV3Certification(candidateLanguage) {
+    if (!candidateLanguage) return;
```
```suggestion
    if (!candidateLanguage) return false;
```

lego-technix replied:
✅ Fait `return false` avec un amend sur le commit concerné.

bpetetot commented on api/src/shared/domain/services/locale-service.js:
```diff
@@ -10,6 +10,8 @@ const CHALLENGE_LOCALES = ['en', 'fr', 'fr-fr', 'nl', 'es', 'it', 'de'];
 
 const SUPPORTED_LOCALES = ['en', 'es', 'fr', 'fr-BE', 'fr-FR', 'nl-BE', 'nl'];
 
+const SUPPORTED_LANGUAGES = ['en', 'es', 'fr', 'nl'];
```
Est-ce qu'on ne pourrait pas générer cette liste depuis `SUPPORTED_LOCALES` en utilisant `Intl.Locale(locale).language` ?

Exemple (code non vérifié):
```js
function getSupportedLanguages() {
  const supportedLanguages = [];
  for (const locale of SUPPORTED_LOCALES) {
   try {
     const { language } = Intl.Locale(locale);
     if (!supportedLanguages.includes(language)) {
       supportedLanguages.push(language);
     }
   } catch {}
  }
  return supportedLanguages;
}
```

lego-technix replied:
Oui on peut tout à fait générer cette liste depuis SUPPORTED_LOCALES en utilisant `Intl.Locale(locale).language`. J’y avais pensé mais je n’avais pas voulu investir de temps là-dessus car une très prochaine étape est de supprimer `getSupportedLanguages` car cette fonction n’est utilisée que pour la création de comptes utilisateur et les utilisateurs n’auront bientôt plus que leur `locale` et plus leur `lang`.

lego-technix replied:
✅ Fait définition de `SUPPORTED_LANGUAGES` à partir de `SUPPORTED_LOCALES`.

PS : Compte tenu que les `SUPPORTED_LOCALES` sont définies par Pix (et pas par des tierces parties comme des utilisateurs), et que `getSupportedLanguages` est testée sur les valeurs définies dans `locale-service`, pas besoin de gérer le `catch`.

bpetetot commented on api/tests/shared/unit/infrastructure/utils/request-response-utils_test.js:
```diff
@@ -66,7 +66,7 @@ describe('Unit | Utils | Request Utils', function () {
       { header: 'fr', expectedLocale: FRENCH_SPOKEN },
       { header: 'en', expectedLocale: ENGLISH_SPOKEN },
-      { header: 'de', expectedLocale: FRENCH_FRANCE },
+      { header: 'tlh', expectedLocale: FRENCH_FRANCE }, // tlh: Klingon locale
```
🤓

lego-technix committed:
refactor(api): derive getSupportedLanguages from SUPPORTED_LOCALES

AndreiaPena commented:
✨ SAV Hera bonjour ✨

Nous avons clôturé puis ré-ouvert votre PR afin de passer sur le fonctionnement Hera.

Les apps Scalingo de votre pull request ont été supprimées automatiquement.
Pour y accéder, cochez les apps que vous souhaitez déployer sur le commentaire de pix-bot-github !

✨ Passez une bonne expérience avec Hera ✨

alexandrecoin commented on api/src/certification/evaluation/application/scenario-simulator-route.js:
```diff
@@ -30,7 +30,7 @@ const register = async (server) => {
               accessibilityAdjustmentNeeded: Joi.boolean(),
               locale: Joi.string()
-                .valid(...Object.values(LOCALE))
+                .valid(...getChallengeLocales())
```
Une remarque pour référence ultérieure mais sans impact apparent sur le fonctionnement (route utilisée uniquement pour de la simulation de déroulé de certification).

Il y a un petit distingo qui s'opère entre ce qui était anciennement renvoyé par la `...Object.values(LOCALE)` -> `['en', 'fr-fr', 'fr', 'nl', 'es']` et `...getChallengeLocales()` -> `['en', 'fr', 'fr-fr', 'nl', 'es', 'it', 'de']`

lego-technix replied:
Oui la liste `getChallengeLocales() -> ['en', 'fr', 'fr-fr', 'nl', 'es', 'it', 'de']` correspond à ce que nous avons trouvé en production actuellement.

Je ne marque pas cette conversation comme résolue pour que ces échanges restent visibles.

lego-technix commented:
J’ai fermé cette PR pour la réouvrir quelques minutes après car la CI ne s’activait pas et la merge queue était vide … mais je n’ai pas pu la réouvrir.

Donc j’ai dû recréer une nouvelle PR identique, c'est la #13085.
